plugins/redfish/fu-plugin-redfish.c

```diff
@@ -414,6 +418,70 @@ fu_plugin_redfish_startup(FuPlugin *plugin, GError **error)
 	return fu_backend_setup(FU_BACKEND(data->backend), error);
 }
 
+static gboolean
+fu_plugin_redfish_cleanup_cb(FuDevice *device, gpointer user_data, GError **error)
+{
+	FuPlugin *self = FU_PLUGIN(user_data);
+	FuPluginData *data = fu_plugin_get_data(self);
+	return fu_backend_setup(FU_BACKEND(data->backend), error);
+}
+
+static gboolean
+fu_plugin_redfish_cleanup(FuPlugin *self,
+			  FuDevice *device,
+			  FuProgress *progress,
+			  FwupdInstallFlags flags,
+			  GError **error)
+{
+	FuPluginData *data = fu_plugin_get_data(self);
+	g_autoptr(FuRedfishRequest) request = fu_redfish_backend_request_new(data->backend);
+	g_autoptr(JsonBuilder) builder = json_builder_new();
+	g_autoptr(GPtrArray) devices = NULL;
+
+	/* nothing to do */
+	if (!fu_device_has_private_flag(device, FU_REDFISH_DEVICE_FLAG_MANAGER_RESET))
+		return TRUE;
+
+	/* ask the BMC to reboot */
+	json_builder_begin_object(builder);
+	json_builder_set_member_name(builder, "ResetType");
+	json_builder_add_string_value(builder, "ForceRestart");
+	json_builder_end_object(builder);
+	if (!fu_redfish_request_perform_full(request,
+					     "/redfish/v1/Managers/1/Actions/Manager.Reset",
+					     "POST",
+					     builder,
+					     FU_REDFISH_REQUEST_PERFORM_FLAG_NONE,
+					     error)) {
+		g_prefix_error(error, "failed to reset manager: ");
+		return FALSE;
+	}
+
+	/* remove all the devices */
+	devices = fu_backend_get_devices(FU_BACKEND(data->backend));
+	for (guint i = 0; i < devices->len; i++) {
+		FuDevice *device_tmp = g_ptr_array_index(devices, i);
+		fu_backend_device_removed(FU_BACKEND(data->backend), device_tmp);
+	}
+
+	/* wait for the BMC to come back */
+	fu_backend_invalidate(FU_BACKEND(data->backend));
+	g_usleep(10 * 1000);
+	if (!fu_device_retry_full(device,
+				  fu_plugin_redfish_cleanup_cb,
+				  FU_REDFISH_PLUGIN_CLEANUP_RETRIES_COUNT,
+				  FU_REDFISH_PLUGIN_CLEANUP_RETRIES_DELAY,
+				  self,
+				  error)) {
+		g_prefix_error(error, "manager failed to come back: ");
+		return FALSE;
+	}
+
+	/* work around an XCC bug... */
```

Comment:
Link to bug upstream?

Reply:
I've pinged Lenovo about some of the reasons for the big delay; there's some wonkyness in the BMC replies too.

Reply:
yeah at least leave some comments on the outcome of this or it will be difficult to remember why +40s.  It might be best to add a plugin quirk for it too.

Reply:
Done and done.